### DISCUSSION TITLE
Support for ORC file format for its union data type

### DISCUSSION THREAD

**Discussion by hinxx** [duckdb/duckdb#6529] [Ideas]
ORC file format supports union data type (https://orc.apache.org/specification/ORCv1/) where only one of the fields holds a value at a time (specified by the tag selector). Parquet has no support for unions and seems like it won't have any time soon (https://github.com/apache/parquet-format/pull/44).

Sadly not even (py)arrow is able to do proper file I/O when it comes to unions (https://github.com/apache/arrow/issues/34262).

Are there  any plans to support ORC file format (similar to how Parquet is supported) in duckDB and with that its union data type, too?

**Comment by hannes** [comment]
No plans, ORC seems far less popular than Parquet

**Comment by osalloum** [comment]
ORC format can provide faster output time on large datasets from Trino, with comparable compression rates to Parquet
+1

**Reply by mahiki** [reply]
Yup, Trino is a main engine at my company and it performs a lot better with ORC files. I have tables that cant be read by DuckDB from s3 because they are in ORC format.

**Comment by caesar168** [comment]
We also want to use duckdb to access ORC format files. Now it's stuck.

**Comment by NicolasPA** [comment]
ORC has been the main file format for big companies still using Hive on Hadoop, and Trino more recently. Most of these remaining big companies are looking into new solutions now, DuckDB could compete better if it supported ORC.

**Comment by hannes** [comment]
Arrow can read ORC it seems, so you could use Arrow's reader to pass ORC file contents into DuckDB as a workaround

**Comment by alitrack** [comment]
There is now a community-built DuckDB ORC extension written in pure Rust: [duckdb_orc](https://github.com/alitrack/duckdb_orc).

It uses `orc-rust` under the hood and supports reading `.orc` files via `read_orc(path)` — including all ORC primitive types, struct, list, map, and all compression codecs (Zlib, Snappy, LZO, LZ4, ZSTD). It also supports glob patterns for multi-file reads and DuckDB's projection pushdown.

I just verified it locally:
- `test.orc` with zstd compression + NULL values → works correctly
- TPC-H `lineitem.orc` (100 rows, 16 columns, dates, doubles, strings) → matches PyArrow results exactly (sum of l_extendedprice = 3918852.53, 100 rows)

Limitations:
- Read-only (no write support yet)
- No filter/predicate pushdown (blocked by missing DuckDB C API — `duckdb_table_function_supports_filter_pushdown` is not exposed to extensions)
- **Union type**: not yet supported by the underlying `orc-rust` reader, so that specific use case from the OP is still a gap

It is not yet submitted to `duckdb/community-extensions` but you can build and `LOAD` it locally today. As an alternative workaround, you can also use PyArrow's ORC reader to pass ORC files into DuckDB via `con.register()`.
